# Investigating WeRateDogs‘s Twitter account
by Emadeldin Fawaz

## Installing needed packages and tweepy

In [2]:
import tweepy
import pandas as pd
import numpy as np
import requests
import json
import time
import datetime as dt
import matplotlib.pyplot as plt

# 1. Data Gathering

In [3]:
# Supplied file
tw_archive = pd.read_csv('twitter-archive-enhanced.csv')

In [5]:
# Requesting Tweet image predictions

with open('image_predictions.tsv' , 'wb') as file:
    image_predictions = requests.get('https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv', auth=('user', 'pass'))
    file.write(image_predictions.content)   

In [6]:
# Reading image predictions
image_predictions = pd.read_csv('image_predictions.tsv', sep='\t')

## Testing the data gathering process so far

In [7]:
tw_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [8]:
tw_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [9]:
image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [10]:
image_predictions.sample(5)

tweet_id                                          jpg_url  \
1511  786233965241827333  https://pbs.twimg.com/media/CulDnZpWcAAGbZ-.jpg   
1603  800141422401830912  https://pbs.twimg.com/media/CxqsX-8XUAAEvjD.jpg   
624   680609293079592961  https://pbs.twimg.com/media/CXICiB9UwAE1sKY.jpg   
1644  808106460588765185  https://pbs.twimg.com/media/Czb4iFRXgAIUMiN.jpg   
225   670338931251150849  https://pbs.twimg.com/media/CU2FsRnVAAA3TEg.jpg   

      img_num                  p1   p1_conf  p1_dog  \
1511        1  Labrador_retriever  0.478193    True   
1603        3    golden_retriever  0.938048    True   
624         1      French_bulldog  0.700764    True   
1644        1    golden_retriever  0.426183    True   
225         1               cairn  0.245033    True   

                               p2   p2_conf  p2_dog  \
1511                   schipperke  0.224817    True   
1603                       kuvasz  0.025119    True   
624                     Chihuahua  0.072390    True   
1644           Labrador_retriever  0.257447    True   
225   West_Highland_white_terrier  0.137709    True   

                                  p3   p3_conf  p3_dog  
1511       Staffordshire_bullterrier  0.077396    True  
1603              Labrador_retriever  0.022977    True  
624   American_Staffordshire_terrier  0.039619    True  
1644                  Great_Pyrenees  0.126482    True  
225              miniature_schnauzer  0.089173    True

In [11]:
# Tweepy keys and token

# CONSUMER_KEY = 'Consumer Key (API key)'
# CONSUMER_SECRET = 'Consumer Secret (API Secret)'
# OAUTH_TOKEN = 'Access Token'
# OAUTH_TOKEN_SECRET = 'Access Token Secret'

consumer_key = 'xxxxconsumer_secret = 'xxxxaccess_token = 'xxxxxaccess_secret = 'xxxxxxx
auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

# Construct the API instance
api = tweepy.API(auth, parser = tweepy.parsers.JSONParser(), wait_on_rate_limit = True, 
                 wait_on_rate_limit_notify = True) 

# Source of the Tweet API Code: Anouar ZBAIDA 

In [16]:
# Liste where we will store the dictionaries of our result
df_list = []
# Liste frame where we will store the tweet_id of the errors
error_list = []

# Calculate the time of excution
start = time.time()

# Get the tweet object for all the teweets in archive dataframe 
for tweet_id in tw_archive['tweet_id']:
    try:
        page = api.get_status(tweet_id, tweet_mode = 'extended')
        # Print one page to look at the structure of the returned file
        # and the names of attributes
        # print(json.dumps(page, indent = 4))
        #break
        
        favorites = page['favorite_count'] # How many favorites the tweet had
        retweets = page['retweet_count'] # Count of the retweet
        user_followers = page['user']['followers_count'] # How many followers the user had
        user_favourites = page['user']['favourites_count'] # How many favorites the user had
        date_time = page['created_at'] # The date and time of the creation
        
        df_list.append({'tweet_id': int(tweet_id),
                        'favorites': int(favorites),
                        'retweets': int(retweets),
                        'user_followers': int(user_followers),
                        'user_favourites': int(user_favourites),
                        'date_time': pd.to_datetime(date_time)})
    
    # Catch the exceptions of the TweepError
    except Exception as e:
        print(str(tweet_id)+ " _ " + str(e))
        error_list.append(tweet_id)

# Calculate the time of excution
end = time.time()
print(end - start)

888202515573088257 _ [{'code': 144, 'message': 'No status found with that ID.'}]
873697596434513921 _ [{'code': 144, 'message': 'No status found with that ID.'}]
872668790621863937 _ [{'code': 144, 'message': 'No status found with that ID.'}]
869988702071779329 _ [{'code': 144, 'message': 'No status found with that ID.'}]
866816280283807744 _ [{'code': 144, 'message': 'No status found with that ID.'}]
861769973181624320 _ [{'code': 144, 'message': 'No status found with that ID.'}]
845459076796616705 _ [{'code': 144, 'message': 'No status found with that ID.'}]
842892208864923648 _ [{'code': 144, 'message': 'No status found with that ID.'}]
837012587749474308 _ [{'code': 144, 'message': 'No status found with that ID.'}]
827228250799742977 _ [{'code': 144, 'message': 'No status found with that ID.'}]
802247111496568832 _ [{'code': 144, 'message': 'No status found with that ID.'}]
Rate limit reached. Sleeping for: 571
775096608509886464 _ [{'code': 144, 'message': 'No status found with th

In [17]:
# We repeat the same operation for the tweet_ids that we coudln't get and append the result to df_list
ee_list = []
for e in error_list:
    try:
        favorites = page['favorite_count']
        retweets = page['retweet_count']
        user_followers = page['user']['followers_count']
        user_favourites = page['user']['favourites_count']
        date_time = page['created_at']
        
        df_list.append({'tweet_id': int(tweet_id),
                        'favorites': int(favorites),
                        'retweets': int(retweets),
                        'user_followers': int(user_followers),
                        'user_favourites': int(user_favourites),
                        'date_time': pd.to_datetime(date_time)})
        
    except Exception:
        print(str(tweet_id)+ " _ " + str(e))
        ee_list.append(e)

In [18]:
# Create DataFrames from list of dictionaries
tweet_status = pd.DataFrame(df_list, columns = ['tweet_id', 'favorites', 'retweets',
                                               'user_followers', 'user_favourites', 'date_time'])

# Save the dataFrame in file
tweet_status.to_csv('tweet_json.txt', encoding = 'utf-8', index=False)

In [21]:
tweet_status = pd.read_csv('tweet_json.txt', encoding = 'utf-8')
tweet_status.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 6 columns):
tweet_id           2356 non-null int64
favorites          2356 non-null int64
retweets           2356 non-null int64
user_followers     2356 non-null int64
user_favourites    2356 non-null int64
date_time          2356 non-null object
dtypes: int64(5), object(1)
memory usage: 110.5+ KB


In [22]:
tweet_status.sample(5)

tweet_id  favorites  retweets  user_followers  \
1237  710997087345876993       4774      1488         7461358   
1515  690248561355657216       1773       438         7461358   
1982  672591762242805761        968       345         7462017   
348   830173239259324417          0      6193         7461337   
1021  744995568523612160       3134       671         7461354   

      user_favourites            date_time  
1237           140021  2016-03-19 01:11:29  
1515           140021  2016-01-21 19:04:15  
1982           140024  2015-12-04 01:42:26  
348            140021  2017-02-10 21:54:58  
1021           140021  2016-06-20 20:49:19

## Data Assessment

In [24]:
# Reading gathered files
tw_archive = pd.read_csv('twitter-archive-enhanced.csv')
image_predictions = pd.read_csv('image_predictions.tsv', sep='\t')
tweet_status = pd.read_csv('tweet_json.txt', encoding = 'utf-8')

### Visual assessment

In [25]:
tw_archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [26]:
image_predictions

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [27]:
tweet_status

tweet_id  favorites  retweets  user_followers  \
0     892420643555336193      38186      8364         7461335   
1     892177421306343426      32756      6176         7461335   
2     891815181378084864      24661      4088         7461335   
3     891689557279858688      41517      8503         7461335   
4     891327558926688256      39707      9201         7461335   
5     891087950875897856      19932      3063         7461335   
6     890971913173991426      11662      2028         7461335   
7     890729181411237888      64434     18557         7461335   
8     890609185150312448      27387      4204         7461335   
9     890240255349198849      31435      7264         7461335   
10    890006608113172480      30224      7215         7461335   
11    889880896479866881      27388      4891         7461335   
12    889665388333682689      47406      9885         7461335   
13    889638837579907072      26759      4473         7461335   
14    889531135344209921      14887      2206         7461335   
15    889278841981685760      24874      5286         7461335   
16    888917238123831296      28675      4420         7461335   
17    888804989199671297      25187      4222         7461335   
18    888554962724278272      19540      3491         7461335   
19    888078434458587136      21444      3441         7461335   
20    887705289381826560      29732      5302         7461335   
21    887517139158093824      45583     11481         7461335   
22    887473957103951883      68062     17878         7461335   
23    887343217045368832      33195     10238         7461335   
24    887101392804085760      30120      5866         7461335   
25    886983233522544640      34641      7644         7461335   
26    886736880519319552      11877      3228         7461335   
27    886680336477933568      22102      4394         7461335   
28    886366144734445568      20877      3147         7461335   
29    886267009285017600        116         4         7461335   
...                  ...        ...       ...             ...   
2326  666099513787052032        153        67         7461403   
2327  666094000022159362        163        72         7461403   
2328  666082916733198337        116        44         7461403   
2329  666073100786774016        318       162         7461403   
2330  666071193221509120        145        59         7461403   
2331  666063827256086533        470       214         7461403   
2332  666058600524156928        111        57         7461403   
2333  666057090499244032        293       140         7461403   
2334  666055525042405380        431       242         7461403   
2335  666051853826850816       1208       841         7461403   
2336  666050758794694657        132        59         7461403   
2337  666049248165822465        108        41         7461403   
2338  666044226329800704        296       139         7461403   
2339  666033412701032449        124        43         7461403   
2340  666029285002620928        129        47         7461403   
2341  666020888022790149       2541       507         7461403   
2342  666020888022790149       2541       507         7461403   
2343  666020888022790149       2541       507         7461403   
2344  666020888022790149       2541       507         7461403   
2345  666020888022790149       2541       507         7461403   
2346  666020888022790149       2541       507         7461403   
2347  666020888022790149       2541       507         7461403   
2348  666020888022790149       2541       507         7461403   
2349  666020888022790149       2541       507         7461403   
2350  666020888022790149       2541       507         7461403   
2351  666020888022790149       2541       507         7461403   
2352  666020888022790149       2541       507         7461403   
2353  666020888022790149       2541       507         7461403   
2354  666020888022790149       2541       507         7461403   
2355  666020888022790149       2541       507      

### Programmatic assessment

### Assessing the Archive document

In [28]:
# Check for anomalies (min or max in comparison to the mnean and the 5 digits description (median and percentiles))
tw_archive.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [29]:
# Check for missing values and if column types are correct as needed
tw_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [55]:
# Check for duplicates
assert [tw_archive.duplicated() == False]

In [30]:
# Find the row(s) where the denominator = 0. This must be a mistake.
tw_archive[tw_archive['rating_denominator'] == 0 ]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
313  835246439529840640           8.352460e+17           26259576.0   

                     timestamp  \
313  2017-02-24 21:54:03 +0000   

                                                source  \
313  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
313  @jonnysun @Lin_Manuel ok jomny I know you're e...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp expanded_urls  \
313                       NaN                        NaN           NaN   

     rating_numerator  rating_denominator  name doggo floofer pupper puppo  
313               960                   0  None  None    None   None  None

In [31]:
# Read the text in the tweet with 0 denominator to evaluate what has gone wrong (if possible)
tw_archive['text'][tw_archive['rating_denominator'] == 0 ]

313    @jonnysun @Lin_Manuel ok jomny I know you're e...
Name: text, dtype: object

Note: delete this row during cleaning.

In [32]:
# Denominators must be 10 (by defacto), check rows with denominators less than 10
tw_archive[tw_archive['rating_denominator'] < 10 ]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
313   835246439529840640           8.352460e+17           26259576.0   
516   810984652412424192                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   

                      timestamp  \
313   2017-02-24 21:54:03 +0000   
516   2016-12-19 23:06:23 +0000   
2335  2015-11-16 16:11:11 +0000   

                                                 source  \
313   <a href="http://twitter.com/download/iphone" r...   
516   <a href="http://twitter.com/download/iphone" r...   
2335  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
313   @jonnysun @Lin_Manuel ok jomny I know you're e...                  NaN   
516   Meet Sam. She smiles 24/7 &amp; secretly aspir...                  NaN   
2335  This is an Albanian 3 1/2 legged  Episcopalian...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
313                        NaN                        NaN   
516                        NaN                        NaN   
2335                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
313                                                 NaN               960   
516   https://www.gofundme.com/sams-smile,https://tw...                24   
2335  https://twitter.com/dog_rates/status/666287406...                 1   

      rating_denominator  name doggo floofer pupper puppo  
313                    0  None  None    None   None  None  
516                    7   Sam  None    None   None  None  
2335                   2    an  None    None   None  None

In [40]:
# Denominators must be 10 (by defacto), check rows with denominators more than 20
tw_archive[tw_archive['rating_denominator'] > 20 ]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
433   820690176645140481                    NaN                  NaN   
902   758467244762497024                    NaN                  NaN   
1120  731156023742988288                    NaN                  NaN   
1202  716439118184652801                    NaN                  NaN   
1228  713900603437621249                    NaN                  NaN   
1254  710658690886586372                    NaN                  NaN   
1274  709198395643068416                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1433  697463031882764288                    NaN                  NaN   
1634  684225744407494656           6.842229e+17         4.196984e+09   
1635  684222868335505415                    NaN                  NaN   
1779  677716515794329600                    NaN                  NaN   
1843  675853064436391936                    NaN                  NaN   

                      timestamp  \
433   2017-01-15 17:52:40 +0000   
902   2016-07-28 01:00:57 +0000   
1120  2016-05-13 16:15:54 +0000   
1202  2016-04-03 01:36:11 +0000   
1228  2016-03-27 01:29:02 +0000   
1254  2016-03-18 02:46:49 +0000   
1274  2016-03-14 02:04:08 +0000   
1351  2016-02-28 21:25:30 +0000   
1433  2016-02-10 16:51:59 +0000   
1634  2016-01-05 04:11:44 +0000   
1635  2016-01-05 04:00:18 +0000   
1779  2015-12-18 05:06:23 +0000   
1843  2015-12-13 01:41:41 +0000   

                                                 source  \
433   <a href="http://twitter.com/download/iphone" r...   
902   <a href="http://twitter.com/download/iphone" r...   
1120  <a href="http://twitter.com/download/iphone" r...   
1202  <a href="http://twitter.com/download/iphone" r...   
1228  <a href="http://twitter.com/download/iphone" r...   
1254  <a href="http://twitter.com/download/iphone" r...   
1274  <a href="http://twitter.com/download/iphone" r...   
1351  <a href="http://twitter.com/download/iphone" r...   
1433  <a href="http://twitter.com/download/iphone" r...   
1634  <a href="http://twitter.com/download/iphone" r...   
1635  <a href="http://twitter.com/download/iphone" r...   
1779  <a href="http://twitter.com/download/iphone" r...   
1843  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
433   The floofs have been released I repeat the flo...                  NaN   
902   Why does this never happen at my front door......                  NaN   
1120  Say hello to this unbelievably well behaved sq...                  NaN   
1202  This is Bluebert. He just saw that both #Final...                  NaN   
1228  Happy Saturday here's 9 puppers on a bench. 99...                  NaN   
1254  Here's a brigade of puppers. All look very pre...                  NaN   
1274  From left to right:\nCletus, Jerome, Alejandro...                  NaN   
1351  Here is a whole flock of puppers.  60/50 I'll ...                  NaN   
1433  Happy Wednesday here's a bucket of pups. 44/40...                  NaN   
1634  Two sneaky puppers were not initially seen, mo...                  NaN   
1635  Someone help the girl is being mugged. Several...                  NaN   
1779  IT'S PUPPERGEDDON. Total of 144/120 ...I think...                  NaN   
1843  Here we have an entire platoon of puppers. Tot...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
433                        NaN                        NaN   
902                        NaN                        NaN   
1120                       NaN                        NaN   
1202                       NaN                        NaN   
1228                       NaN                        NaN   
1254                       NaN                        NaN   
1274                       NaN                        NaN   
1351                       NaN                        NaN   
1433               

In [33]:
# Numerators must be more than 10 (this is what is unique about this WeRateDog). Check rows with 0 values in numerator
tw_archive[tw_archive['rating_numerator'] == 0 ]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
315   835152434251116546                    NaN                  NaN   
1016  746906459439529985           7.468859e+17         4.196984e+09   

                      timestamp  \
315   2017-02-24 15:40:31 +0000   
1016  2016-06-26 03:22:31 +0000   

                                                 source  \
315   <a href="http://twitter.com/download/iphone" r...   
1016  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
315   When you're so blinded by your systematic plag...                  NaN   
1016  PUPDATE: can't see any. Even if I could, I cou...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
315                        NaN                        NaN   
1016                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
315   https://twitter.com/dog_rates/status/835152434...                 0   
1016  https://twitter.com/dog_rates/status/746906459...                 0   

      rating_denominator  name doggo floofer pupper puppo  
315                   10  None  None    None   None  None  
1016                  10  None  None    None   None  None

In [34]:
# Double check that it is only two rows
len(tw_archive[tw_archive['rating_numerator'] == 0 ])

2

In [35]:
# Check anomalies in the numerator (very large values or values way more than the average)
tw_archive[tw_archive['rating_numerator'] > 20]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
188   855862651834028034           8.558616e+17         1.943518e+08   
189   855860136149123072           8.558585e+17         1.361572e+07   
290   838150277551247360           8.381455e+17         2.195506e+07   
313   835246439529840640           8.352460e+17         2.625958e+07   
340   832215909146226688                    NaN                  NaN   
433   820690176645140481                    NaN                  NaN   
516   810984652412424192                    NaN                  NaN   
695   786709082849828864                    NaN                  NaN   
763   778027034220126208                    NaN                  NaN   
902   758467244762497024                    NaN                  NaN   
979   749981277374128128                    NaN                  NaN   
1120  731156023742988288                    NaN                  NaN   
1202  716439118184652801                    NaN                  NaN   
1228  713900603437621249                    NaN                  NaN   
1254  710658690886586372                    NaN                  NaN   
1274  709198395643068416                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1433  697463031882764288                    NaN                  NaN   
1634  684225744407494656           6.842229e+17         4.196984e+09   
1635  684222868335505415                    NaN                  NaN   
1712  680494726643068929                    NaN                  NaN   
1779  677716515794329600                    NaN                  NaN   
1843  675853064436391936                    NaN                  NaN   
2074  670842764863651840                    NaN                  NaN   

                      timestamp  \
188   2017-04-22 19:15:32 +0000   
189   2017-04-22 19:05:32 +0000   
290   2017-03-04 22:12:52 +0000   
313   2017-02-24 21:54:03 +0000   
340   2017-02-16 13:11:49 +0000   
433   2017-01-15 17:52:40 +0000   
516   2016-12-19 23:06:23 +0000   
695   2016-10-13 23:23:56 +0000   
763   2016-09-20 00:24:34 +0000   
902   2016-07-28 01:00:57 +0000   
979   2016-07-04 15:00:45 +0000   
1120  2016-05-13 16:15:54 +0000   
1202  2016-04-03 01:36:11 +0000   
1228  2016-03-27 01:29:02 +0000   
1254  2016-03-18 02:46:49 +0000   
1274  2016-03-14 02:04:08 +0000   
1351  2016-02-28 21:25:30 +0000   
1433  2016-02-10 16:51:59 +0000   
1634  2016-01-05 04:11:44 +0000   
1635  2016-01-05 04:00:18 +0000   
1712  2015-12-25 21:06:00 +0000   
1779  2015-12-18 05:06:23 +0000   
1843  2015-12-13 01:41:41 +0000   
2074  2015-11-29 05:52:33 +0000   

                                                 source  \
188   <a href="http://twitter.com/download/iphone" r...   
189   <a href="http://twitter.com/download/iphone" r...   
290   <a href="http://twitter.com/download/iphone" r...   
313   <a href="http://twitter.com/download/iphone" r...   
340   <a href="http://twitter.com/download/iphone" r...   
433   <a href="http://twitter.com/download/iphone" r...   
516   <a href="http://twitter.com/download/iphone" r...   
695   <a href="http://twitter.com/download/iphone" r...   
763   <a href="http://twitter.com/download/iphone" r...   
902   <a href="http://twitter.com/download/iphone" r...   
979   <a href="https://about.twitter.com/products/tw...   
1120  <a href="http://twitter.com/download/iphone" r...   
1202  <a href="http://twitter.com/download/iphone" r...   
1228  <a href="http://twitter.com/download/iphone" r...   
1254  <a href="http://twitter.com/download/iphone" r...   
1274  <a href="http://twitter.com/download/iphone" r...   
1351  <a href="http://twitter.com/download/iphone" r...   
1433  <a href="http://twitter.com/download/iphone" r...   
1634  <a href="http://twitter.com/download/iphone" r...   
1635  <a href="http://twitter.com/download/iphone" r...   
1712  <a href="http://twitter.com/download/iphone" r...   
1779  <a href="http://twitter.com/downloa

In [36]:
# Count rows with high numerator (more than 20)
len(tw_archive[tw_archive['rating_numerator'] > 20 ])

24

There are 24 rows with anomalies in the numerator (>20). Deleting these rows will enhance the statestics (such as median) and data visualization (by inimizing outliers). However, before deleting all of them, check if the values are wrong or could be fixed.


As the Denominator is fixed to be 10, capturing strange values like 40, 50 or 80 indicates a mistake in data entry (probably by the Twitter Account holder him/herself). This might be even more illustrated when the numerator also has the value of 44, 40, 50, 80, 88. This indicates that a typo mistake happened here where a 4, 5 or 8 was mixed with an 1.

### Assissing the Image Document

In [37]:
# Check for missing values and column types
image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [38]:
# Check for anomalies and the basic descriptive statistics
image_predictions.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [56]:
# Check fr duplicates
assert [image_predictions.duplicated() == False]

In [44]:
# Check the values in the image predictions and if there are any an usual value
image_predictions.p1.value_counts()

golden_retriever             150
Labrador_retriever           100
Pembroke                      89
Chihuahua                     83
pug                           57
chow                          44
Samoyed                       43
toy_poodle                    39
Pomeranian                    38
cocker_spaniel                30
malamute                      30
French_bulldog                26
Chesapeake_Bay_retriever      23
miniature_pinscher            23
seat_belt                     22
German_shepherd               20
Siberian_husky                20
Staffordshire_bullterrier     20
web_site                      19
Cardigan                      19
Maltese_dog                   18
teddy                         18
beagle                        18
Eskimo_dog                    18
Shetland_sheepdog             18
Lakeland_terrier              17
Rottweiler                    17
Shih-Tzu                      17
kuvasz                        16
Italian_greyhound             16
          

In [45]:
# Find how many rare things (frequency less than 3) appeared in the data and investigaate what they are
rare_things= image_predictions.groupby('p1').filter(lambda x: len(x) < 3)

In [46]:
rare_things.sample(5)

tweet_id                                          jpg_url  \
28    666407126856765440  https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
1970  868880397819494401  https://pbs.twimg.com/media/DA7iHL5U0AA1OQo.jpg   
929   702932127499816960  https://pbs.twimg.com/media/CcFRCfRW4AA5a72.jpg   
1856  841439858740625411  https://pbs.twimg.com/media/C61lFFiWoAAJdiL.jpg   
1755  824796380199809024  https://pbs.twimg.com/media/CwiuEJmW8AAZnit.jpg   

      img_num                       p1   p1_conf  p1_dog                  p2  \
28          1  black-and-tan_coonhound  0.529139    True          bloodhound   
1970        1                   laptop  0.153718   False      French_bulldog   
929         1                  wallaby  0.410710   False              wombat   
1856        3         military_uniform  0.853684   False  Labrador_retriever   
1755        2                 gas_pump  0.676439   False           harvester   

       p2_conf  p2_dog                     p3   p3_conf  p3_dog  
28    0.244220    True  flat-coated_retriever  0.173810    True  
1970  0.099984    True                printer  0.077130   False  
929   0.239332   False                 beaver  0.149605   False  
1856  0.048200    True            groenendael  0.015394    True  
1755  0.049995   False                  swing  0.044660   False

In [47]:
len(rare_things)

271

In [16]:
len(image_predictions)

2075

### Assessing Tweet Status


In [58]:
# Check for missing values and column types
tweet_status.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 6 columns):
tweet_id           2356 non-null int64
favorites          2356 non-null int64
retweets           2356 non-null int64
user_followers     2356 non-null int64
user_favourites    2356 non-null int64
date_time          2356 non-null object
dtypes: int64(5), object(1)
memory usage: 110.5+ KB


In [60]:
# Check for duplicates
assert [tweet_status.tweet_id.duplicated() == False]

In [61]:
# Check for anomalies
tweet_status.describe()

tweet_id      favorites      retweets  user_followers  \
count  2.356000e+03    2356.000000   2356.000000    2.356000e+03   
mean   7.417684e+17    7962.654499   2933.644737    7.461371e+06   
std    6.837209e+16   12345.699353   4947.190933    6.167252e+01   
min    6.660209e+17       0.000000      0.000000    7.461335e+06   
25%    6.776996e+17    1398.500000    583.000000    7.461340e+06   
50%    7.178159e+17    3461.500000   1365.000000    7.461357e+06   
75%    7.986755e+17    9779.750000   3417.250000    7.461400e+06   
max    8.924206e+17  164825.000000  84021.000000    7.462017e+06   

       user_favourites  
count      2356.000000  
mean     140021.024194  
std           0.268376  
min      140021.000000  
25%      140021.000000  
50%      140021.000000  
75%      140021.000000  
max      140024.000000

### Assessment report

<strong>Tidiness Issues (structural issues)</strong>

<strong>Tw_archive</strong>

- Not all columns are needed for further analysis

- Replace 'None' with np.nan to indicate the missing values

-	The columns (doggo, floofer, pupper and puppo) do not need to be separated. Each dog will be classified as one of these classifications. It is better to create one column for dog classification that contains the values (doggo, floofer, pupper and puppo).



<b>image_predictions</b>

-   Change the names of columns type (p1, p2 and p3) to a more descriptive name. 
-	The P1 confidence interval might not be always convincing. I will set the acceptance threshold at P ≥ 0.5. This might mean neglecting a lot of the data (the median is .585). However, the trade off is higher confidence in the results.


<b>tweet_status</b>

- The two columns user_followers & user_favourites contain fixed values and will not be needed in the analysis.




<strong>Quality issues</strong>

<b>Tw_archive</b>

-	Some columns like (in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id) are missing too many values and might not be needed in this analysis.
-	For the column “rating_numirator” there are zero values. delete these two rows with zero in the numirator.
-	For the column “rating_numirator” there are extreme values. 
-	Delete the cells with numerator more than 20. This will enhance visualization (by reducing outliers) and statistics. 
-	For the column “rating_denumirator” there is a zero value. delete it. 
-	For the column “rating_denumirator” there are extreme values. 
-	Correcting anomalies and typo mistakes in the numerator and the denominator:
As the Denominator is fixed to be 10, capturing strange values like 40, 50 or 80 indicates a mistake in data entry (probably by the Twitter Account holder him/herself). This might be even more illustrated when the numerator also has the value of 44, 40, 50, 80, 88. This indicates that a typo mistake happened here where a 4, 5 or 8 was mixed with an 1.
-	Line  1202‎	numerator must be 10 not 50, and denominator must be 10 (typo 5 instead of 1)‎
-	Line  1254‎	numerator must be 10 not 80, and denominator must be 10 (typo 8 ‎instead of 1)‎
-	Line  1433‎	numerator must be 11 not 44, and denominator must be 10 (typo 4 ‎instead of 1)‎
-	Line  1843‎	numerator must be 11 not 88, and denominator must be 10 (typo 8 ‎instead of 1)‎
-	It is not clear which typo mistake is it with the rest of the extreme denominator ‎values.‎



<b>image_predictions</b>

- Drop unneeded columns
- “Egyptian_cat”, is that a dog? Delete this row and rows with similar strange values like for eg. "Laptop".
- There are some strange values in the p1 columns such as “school_bus, pillow, cartoon” ‎these cells must be checked first then deleted if not needed. 
- 271 rows (with object frequency less than 3) could be deleted because they are not that frequent and most probably not dogs. This will solve the problems in the previous two points

<b>tweet_status</b>

-	Drop unneeded columns



<strong>Mereg all documents usig tweet_id</strong>

## Data Cleaning

## Cleaning tw_archive

In [62]:
# Reading gathered files
tw_archive = pd.read_csv('twitter-archive-enhanced.csv')
image_predictions = pd.read_csv('image_predictions.tsv', sep='\t')
tweet_status = pd.read_csv('tweet_json.txt', encoding = 'utf-8')

In [63]:
# create a cleaned dataframe from the archive
cleaned_archive = tw_archive.copy()

In [64]:
# test
cleaned_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

The column "retweeted_status_id" is showing if the tweet was original or a reply to another tweet.
Select only rows where this column has a NaN value. This should reveal the original tweets and delete all retweets.


In [65]:
# Define: Select rows wehre the column "in_reply_to_status_id" is Null (this will exclude all the retweets)

#Code
cleaned_archive = cleaned_archive[cleaned_archive['retweeted_status_id'].isnull()]

In [66]:
# test
cleaned_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2175 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2175 non-null object
source                        2175 non-null object
text                          2175 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2117 non-null object
rating_numerator              2175 non-null int64
rating_denominator            2175 non-null int64
name                          2175 non-null object
doggo                         2175 non-null object
floofer                       2175 non-null object
pupper                        2175 non-null object
puppo                         2175 non-null object
dtypes: float64(4), int64(3), object(1

In [67]:
# Define: Dropping unneded (missing too much values) columns using drop function.

# code
cleaned_archive = cleaned_archive.drop(['in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp'], axis=1)

In [68]:
# Test
cleaned_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2175 non-null int64
timestamp             2175 non-null object
source                2175 non-null object
text                  2175 non-null object
expanded_urls         2117 non-null object
rating_numerator      2175 non-null int64
rating_denominator    2175 non-null int64
name                  2175 non-null object
doggo                 2175 non-null object
floofer               2175 non-null object
pupper                2175 non-null object
puppo                 2175 non-null object
dtypes: int64(3), object(9)
memory usage: 220.9+ KB


In [69]:
# Define: I do not need 'expanded_urls' in my analysis. use drop function
# Code
cleaned_archive = cleaned_archive.drop(['expanded_urls'], axis = 1)

In [70]:
# Test
cleaned_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 11 columns):
tweet_id              2175 non-null int64
timestamp             2175 non-null object
source                2175 non-null object
text                  2175 non-null object
rating_numerator      2175 non-null int64
rating_denominator    2175 non-null int64
name                  2175 non-null object
doggo                 2175 non-null object
floofer               2175 non-null object
pupper                2175 non-null object
puppo                 2175 non-null object
dtypes: int64(3), object(8)
memory usage: 203.9+ KB


In [75]:
# Define: use loc to correct the typo mistakes in specified cells

# Code

cleaned_archive.loc[cleaned_archive['rating_numerator']==50, ['rating_numerator']] = 10
cleaned_archive.loc[cleaned_archive['rating_denominator']==50, ['rating_denominator']] = 10

cleaned_archive.loc[cleaned_archive['rating_numerator']==88, ['rating_numerator']] = 11
cleaned_archive.loc[cleaned_archive['rating_denominator']==80, ['rating_denominator']] = 10

cleaned_archive.loc[cleaned_archive['rating_numerator']==80, ['rating_numerator']] = 10
cleaned_archive.loc[cleaned_archive['rating_denominator']==80, ['rating_denominator']] = 10

cleaned_archive.loc[cleaned_archive['rating_numerator']==44, ['rating_numerator']] = 11
cleaned_archive.loc[cleaned_archive['rating_denominator']==40, ['rating_denominator']] = 10



In [76]:
# test
cleaned_archive[cleaned_archive['rating_numerator'] > 20 ]

tweet_id                  timestamp  \
188   855862651834028034  2017-04-22 19:15:32 +0000   
189   855860136149123072  2017-04-22 19:05:32 +0000   
290   838150277551247360  2017-03-04 22:12:52 +0000   
313   835246439529840640  2017-02-24 21:54:03 +0000   
433   820690176645140481  2017-01-15 17:52:40 +0000   
516   810984652412424192  2016-12-19 23:06:23 +0000   
695   786709082849828864  2016-10-13 23:23:56 +0000   
763   778027034220126208  2016-09-20 00:24:34 +0000   
902   758467244762497024  2016-07-28 01:00:57 +0000   
979   749981277374128128  2016-07-04 15:00:45 +0000   
1120  731156023742988288  2016-05-13 16:15:54 +0000   
1228  713900603437621249  2016-03-27 01:29:02 +0000   
1274  709198395643068416  2016-03-14 02:04:08 +0000   
1351  704054845121142784  2016-02-28 21:25:30 +0000   
1634  684225744407494656  2016-01-05 04:11:44 +0000   
1635  684222868335505415  2016-01-05 04:00:18 +0000   
1712  680494726643068929  2015-12-25 21:06:00 +0000   
1779  677716515794329600  2015-12-18 05:06:23 +0000   
2074  670842764863651840  2015-11-29 05:52:33 +0000   

                                                 source  \
188   <a href="http://twitter.com/download/iphone" r...   
189   <a href="http://twitter.com/download/iphone" r...   
290   <a href="http://twitter.com/download/iphone" r...   
313   <a href="http://twitter.com/download/iphone" r...   
433   <a href="http://twitter.com/download/iphone" r...   
516   <a href="http://twitter.com/download/iphone" r...   
695   <a href="http://twitter.com/download/iphone" r...   
763   <a href="http://twitter.com/download/iphone" r...   
902   <a href="http://twitter.com/download/iphone" r...   
979   <a href="https://about.twitter.com/products/tw...   
1120  <a href="http://twitter.com/download/iphone" r...   
1228  <a href="http://twitter.com/download/iphone" r...   
1274  <a href="http://twitter.com/download/iphone" r...   
1351  <a href="http://twitter.com/download/iphone" r...   
1634  <a href="http://twitter.com/download/iphone" r...   
1635  <a href="http://twitter.com/download/iphone" r...   
1712  <a href="http://twitter.com/download/iphone" r...   
1779  <a href="http://twitter.com/download/iphone" r...   
2074  <a href="http://twitter.com/download/iphone" r...   

                                                   text  rating_numerator  \
188   @dhmontgomery We also gave snoop dogg a 420/10...               420   
189   @s8n You tried very hard to portray this good ...               666   
290                                  @markhoppus 182/10               182   
313   @jonnysun @Lin_Manuel ok jomny I know you're e...               960   
433   The floofs have been released I repeat the flo...                84   
516   Meet Sam. She smiles 24/7 &amp; secretly aspir...                24   
695   This is Logan, the Chow who lived. He solemnly...                75   
763   This is Sophie. She's a Jubilant Bush Pupper. ...                27   
902   Why does this never happen at my front door......               165   
979   This is Atticus. He's quite simply America af....              1776   
1120  Say hello to this unbelievably well behaved sq...               204   
1228  Happy Saturday here's 9 puppers on a bench. 99...                99   
1274  From left to right:\nCletus, Jerome, Alejandro...                45   
1351  Here is a whole flock of puppers.  60/50 I'll ...                60   
1634  Two sneaky puppers were not initially seen, mo...               143   
1635  Someone help the girl is being mugged. Several...               121   
1712  Here we have uncovered an entire battalion of ...                26   
1779  IT'S PUPPERGEDDON. Total of 144/120 ...I think...               144   
2074  After so many requests... here you go.\n\nGood...               420   

      rating_denominator     name doggo floofer  pupper puppo  
188                   10     None  None    None    None  None  
189                   10     None  None    None    None  None  
2

In [77]:
len(cleaned_archive[cleaned_archive['rating_numerator'] > 20 ])

19

In [78]:
cleaned_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 11 columns):
tweet_id              2175 non-null int64
timestamp             2175 non-null object
source                2175 non-null object
text                  2175 non-null object
rating_numerator      2175 non-null int64
rating_denominator    2175 non-null int64
name                  2175 non-null object
doggo                 2175 non-null object
floofer               2175 non-null object
pupper                2175 non-null object
puppo                 2175 non-null object
dtypes: int64(3), object(8)
memory usage: 203.9+ KB


In [79]:
# Define: exclude all extreme and zero values from the numertor and denuminator

#Code
cleaned_archive = cleaned_archive[cleaned_archive['rating_numerator'] != 0 ]
cleaned_archive = cleaned_archive[cleaned_archive['rating_denominator'] >= 10 ]
cleaned_archive = cleaned_archive[cleaned_archive['rating_numerator'] <= 20 ]
cleaned_archive = cleaned_archive[cleaned_archive['rating_denominator'] < 20 ]


In [80]:
# test
len(cleaned_archive[cleaned_archive['rating_numerator'] > 20 ])

0

In [81]:
len(cleaned_archive[cleaned_archive['rating_denominator'] < 10 ])

0

In [82]:
len(cleaned_archive[cleaned_archive['rating_denominator'] > 20 ])

0

In [83]:
cleaned_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2151 entries, 0 to 2355
Data columns (total 11 columns):
tweet_id              2151 non-null int64
timestamp             2151 non-null object
source                2151 non-null object
text                  2151 non-null object
rating_numerator      2151 non-null int64
rating_denominator    2151 non-null int64
name                  2151 non-null object
doggo                 2151 non-null object
floofer               2151 non-null object
pupper                2151 non-null object
puppo                 2151 non-null object
dtypes: int64(3), object(8)
memory usage: 201.7+ KB


In [84]:
# Define: replace the value 'None' with the NaN (missing value)

# Code:
cleaned_archive =  cleaned_archive.replace('None', np.nan)

In [85]:
# test
cleaned_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2151 entries, 0 to 2355
Data columns (total 11 columns):
tweet_id              2151 non-null int64
timestamp             2151 non-null object
source                2151 non-null object
text                  2151 non-null object
rating_numerator      2151 non-null int64
rating_denominator    2151 non-null int64
name                  1488 non-null object
doggo                 87 non-null object
floofer               10 non-null object
pupper                233 non-null object
puppo                 25 non-null object
dtypes: int64(3), object(8)
memory usage: 201.7+ KB


In [86]:
# Define: use loc to add a new column dog_class = doggo, floofer, pupper or poppo. NaN will be used if not any of the previous

# Code
cleaned_archive.loc[cleaned_archive['doggo'] == 'doggo', 'dog_class'] = 'doggo'
cleaned_archive.loc[cleaned_archive['floofer'] == 'floofer', 'dog_class'] = 'floofer'
cleaned_archive.loc[cleaned_archive['pupper'] == 'pupper', 'dog_class'] = 'pupper'
cleaned_archive.loc[cleaned_archive['puppo'] == 'puppo', 'dog_class'] = 'puppo'



In [88]:
# Test
cleaned_archive.sample(10)

tweet_id                  timestamp  \
1251  710997087345876993  2016-03-19 01:11:29 +0000   
1655  683391852557561860  2016-01-02 20:58:09 +0000   
175   857989990357356544  2017-04-28 16:08:49 +0000   
1037  744709971296780288  2016-06-20 01:54:27 +0000   
283   838952994649550848  2017-03-07 03:22:35 +0000   
2323  666428276349472768  2015-11-17 01:30:57 +0000   
377   828011680017821696  2017-02-04 22:45:42 +0000   
1567  687826841265172480  2016-01-15 02:41:12 +0000   
2154  669625907762618368  2015-11-25 21:17:12 +0000   
1578  687317306314240000  2016-01-13 16:56:30 +0000   

                                                 source  \
1251  <a href="http://twitter.com/download/iphone" r...   
1655  <a href="http://twitter.com/download/iphone" r...   
175   <a href="http://twitter.com/download/iphone" r...   
1037  <a href="http://twitter.com/download/iphone" r...   
283   <a href="http://twitter.com/download/iphone" r...   
2323  <a href="http://twitter.com/download/iphone" r...   
377   <a href="http://twitter.com/download/iphone" r...   
1567  <a href="http://twitter.com/download/iphone" r...   
2154  <a href="http://twitter.com/download/iphone" r...   
1578  <a href="http://twitter.com/download/iphone" r...   

                                                   text  rating_numerator  \
1251  Meet Milo and Amos. They are the best of pals....                12   
1655  Say hello to Jack (pronounced "Kevin"). He's a...                11   
175   This is Rosie. She was just informed of the wa...                12   
1037  This is Harvey. He's stealthy af. 10/10 would ...                10   
283   SHE MISPLACED HER HOOMAN 13/10 MISTAKES HAPPEN...                13   
2323  Here we have an Austrian Pulitzer. Collectors ...                 7   
377   Say hello to Brutus and Jersey. They think the...                11   
1567  This is Louis. He's a rollercoaster of emotion...                12   
2154  This is Penny. She's having fun AND being safe...                12   
1578  This is Tyrone. He's a leaf wizard. Self-motiv...                11   

      rating_denominator    name doggo floofer pupper puppo dog_class  
1251                  10    Milo   NaN     NaN    NaN   NaN       NaN  
1655                  10    Jack   NaN     NaN    NaN   NaN       NaN  
175                   10   Rosie   NaN     NaN    NaN   NaN       NaN  
1037                  10  Harvey   NaN     NaN    NaN   NaN       NaN  
283                   10     NaN   NaN     NaN    NaN   NaN       NaN  
2323                  10     NaN   NaN     NaN    NaN   NaN       NaN  
377                   10  Brutus   NaN     NaN    NaN   NaN       NaN  
1567                  10   Louis   NaN     NaN    NaN   NaN       NaN  
2154                  10   Penny   NaN     NaN    NaN   NaN       NaN  
1578                  10  Tyrone   NaN     NaN    NaN   NaN       NaN

In [89]:
# test
cleaned_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2151 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2151 non-null int64
timestamp             2151 non-null object
source                2151 non-null object
text                  2151 non-null object
rating_numerator      2151 non-null int64
rating_denominator    2151 non-null int64
name                  1488 non-null object
doggo                 87 non-null object
floofer               10 non-null object
pupper                233 non-null object
puppo                 25 non-null object
dog_class             343 non-null object
dtypes: int64(3), object(9)
memory usage: 218.5+ KB


In [90]:
# dropping unneded doggo, floofer, pupper or poppo columns
cleaned_archive = cleaned_archive.drop(['doggo', 'floofer', 'pupper', 'puppo'], axis = 1)

In [91]:
cleaned_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2151 entries, 0 to 2355
Data columns (total 8 columns):
tweet_id              2151 non-null int64
timestamp             2151 non-null object
source                2151 non-null object
text                  2151 non-null object
rating_numerator      2151 non-null int64
rating_denominator    2151 non-null int64
name                  1488 non-null object
dog_class             343 non-null object
dtypes: int64(3), object(5)
memory usage: 151.2+ KB


### Cleaning images dataframe

In [92]:
# Create a working df 
cleaned_images = image_predictions.copy()

In [93]:
# Test
cleaned_images.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [94]:
# Define: rename columns 

# Code
cleaned_images= cleaned_images.rename(columns={'p1':'Breed_probability1', 'p2':'Breed_probability2', 'p3':'Breed_probability3'})


In [95]:
# Test
cleaned_images.sample(5)

tweet_id                                          jpg_url  \
514   676237365392908289  https://pbs.twimg.com/media/CWJ6Sc-WwAAlpI6.jpg   
1922  856543823941562368  https://pbs.twimg.com/media/C-MOEDCXYAEjp7o.jpg   
1087  718939241951195136  https://pbs.twimg.com/media/CfovbK4WIAAkTn3.jpg   
1365  761599872357261312  https://pbs.twimg.com/media/CpG_CrlWYAYyuP3.jpg   
818   692901601640583168  https://pbs.twimg.com/media/CZ2uU37UcAANzmK.jpg   

      img_num           Breed_probability1   p1_conf  p1_dog  \
514         1               French_bulldog  0.961996    True   
1922        1                  Boston_bull  0.306910    True   
1087        1                     Pembroke  0.766327    True   
1365        1                Gordon_setter  0.240427    True   
818         1  soft-coated_wheaten_terrier  0.403496    True   

     Breed_probability2   p2_conf  p2_dog Breed_probability3   p3_conf  p3_dog  
514           Chihuahua  0.021793    True        Boston_bull  0.006916    True  
1922        Siamese_cat  0.191218   False          Chihuahua  0.189288    True  
1087           Cardigan  0.222126    True      toilet_tissue  0.006757   False  
1365             Saluki  0.224269    True           Doberman  0.129730    True  
818      cocker_spaniel  0.135164    True   golden_retriever  0.088719    True

In [96]:
# Define: Exclude all rows with P1_confidence less than 0.5

# Code: 
cleaned_images = cleaned_images[cleaned_images['p1_conf'] > 0.5 ]

In [97]:
# test
assert [cleaned_images[cleaned_images['p1_conf'] > 0.5 ]]

In [98]:
# Define: use filter to select only needed columns

# Code
cleaned_images = cleaned_images.filter(['tweet_id','Breed_probability1','p1_conf'] )



In [99]:
# test 
cleaned_images.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1239 entries, 1 to 2072
Data columns (total 3 columns):
tweet_id              1239 non-null int64
Breed_probability1    1239 non-null object
p1_conf               1239 non-null float64
dtypes: float64(1), int64(1), object(1)
memory usage: 38.7+ KB


In [100]:
# Define: use groupby to select not rare things in the data (frequency more than 3)

# Code
cleaned_images = cleaned_images.groupby('Breed_probability1').filter(lambda x: len(x) > 3)

In [101]:
# Test
cleaned_images.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 987 entries, 2 to 2072
Data columns (total 3 columns):
tweet_id              987 non-null int64
Breed_probability1    987 non-null object
p1_conf               987 non-null float64
dtypes: float64(1), int64(1), object(1)
memory usage: 30.8+ KB


## Cleaning tweet_status

In [102]:
#Create a working df
cleaned_status = tweet_status.copy()

In [103]:
cleaned_status.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 6 columns):
tweet_id           2356 non-null int64
favorites          2356 non-null int64
retweets           2356 non-null int64
user_followers     2356 non-null int64
user_favourites    2356 non-null int64
date_time          2356 non-null object
dtypes: int64(5), object(1)
memory usage: 110.5+ KB


In [104]:
# Define: filter/select needed columns
# Code
cleaned_status = cleaned_status.filter(['tweet_id','favorites','retweets', 'date_time'] )

In [105]:
# test
cleaned_status.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 4 columns):
tweet_id     2356 non-null int64
favorites    2356 non-null int64
retweets     2356 non-null int64
date_time    2356 non-null object
dtypes: int64(3), object(1)
memory usage: 73.7+ KB


## Merging documents to form a working dataframe

In [106]:
tweeter_df = pd.merge(cleaned_archive, cleaned_images, how='outer', on=['tweet_id'])

In [107]:
tweeter_df = pd.merge(tweeter_df, cleaned_status, how = 'outer', on=['tweet_id'])

In [108]:
# Test the merge
tweeter_df.sample(5)

tweet_id                  timestamp  \
2072  667443425659232256  2015-11-19 20:44:47 +0000   
839   745789745784041472  2016-06-23 01:25:06 +0000   
882   739979191639244800  2016-06-07 00:36:02 +0000   
574   783334639985389568  2016-10-04 15:55:06 +0000   
59    880095782870896641  2017-06-28 16:09:20 +0000   

                                                 source  \
2072  <a href="http://twitter.com/download/iphone" r...   
839   <a href="http://twitter.com/download/iphone" r...   
882   <a href="http://twitter.com/download/iphone" r...   
574   <a href="http://twitter.com/download/iphone" r...   
59    <a href="http://twitter.com/download/iphone" r...   

                                                   text  rating_numerator  \
2072  Exotic dog here. Long neck. Weird paws. Obsess...               6.0   
839   This is Gus. He didn't win the Powerball. Quit...              10.0   
882   This is Nollie. She's waving at you. If you do...              12.0   
574   This is Dave. He's currently in a predicament....              12.0   
59    Please don't send in photos without dogs in th...              11.0   

      rating_denominator    name dog_class Breed_probability1   p1_conf  \
2072                10.0     NaN       NaN                NaN       NaN   
839                 10.0     Gus       NaN           Pekinese  0.984267   
882                 10.0  Nollie       NaN                NaN       NaN   
574                 10.0    Dave       NaN           Cardigan  0.593858   
59                  10.0     NaN       NaN                NaN       NaN   

      favorites  retweets            date_time  
2072      791.0     591.0  2015-11-19 20:44:47  
839      4249.0    1147.0  2016-06-23 01:25:06  
882     21009.0    6347.0  2016-06-07 00:36:02  
574     31366.0   12902.0  2016-10-04 15:55:06  
59      27265.0    4297.0  2017-06-28 16:09:20

In [109]:
tweeter_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2360 entries, 0 to 2359
Data columns (total 13 columns):
tweet_id              2360 non-null int64
timestamp             2165 non-null object
source                2165 non-null object
text                  2165 non-null object
rating_numerator      2165 non-null float64
rating_denominator    2165 non-null float64
name                  1488 non-null object
dog_class             343 non-null object
Breed_probability1    987 non-null object
p1_conf               987 non-null float64
favorites             2356 non-null float64
retweets              2356 non-null float64
date_time             2356 non-null object
dtypes: float64(5), int64(1), object(7)
memory usage: 258.1+ KB


There are some missing values in some columns. However, based on the analysis I want to conduct this dataframe is good to go.
I do not want to drop information now that I might need later.

In [110]:
# Saving df as csv
tweeter_df.to_csv('twitter_archive_master.csv')